### PR TITLE
Augment "chunkSize" to metadata when declaring new metrics

### LINKED ISSUE #65: Add chunk size to metadata
- [ ] Add `chunkSize` https://metricq.github.io/metricq-python/metadata.html#standardized-fields - `null` is not known
- [ ] Infer the value during `declare_metrics` from `Source.chunk_size`

### PR DESCRIPTION
**Pull request by phijor**
This fixes #65 by adding "chunkSize" to the metadata of each metric before sending the `source.declare_metrics` RPC.

Technically, this introduces a **breaking change** (so I marked it as 2.0.0):

* 4dc2190: `chunk_size` is validated before assignment, that may raise an exception. Assigning `0` instead of `None` is now an error, where before it was silently converted

Some minor issues that I originally found here moved to #74.

### CODE REVIEW

File: metricq/source.py [hunk 1]
```diff
@@ -89,31 +89,26 @@ async def task(self):
...
 
-        To reduce network and packet overhead, it may be advisable to send multiple data points at once.
-        Be aware that there is an `overhead-latency trade-off` to be made:
-        If your Source produces one data point every :math:`10` seconds, having a :code:`chunk_size` of :code:`10` means that it takes almost :math:`2` minutes (:math:`100` s) before a chunk is is sent.
-        If instead it produces :math:`1000` data points per second, network load can be reduced by setting a value of :code:`1000` without affecting latency too much.
-        """
-        return None if self._chunk_size == 0 else self._chunk_size
+    Raises:
+        TypeError: if value set is neither :literal:`None` nor an integer
+        ValueError: if value set is not a positive integer
```
Comment (bmario) [inline metricq/source.py:105]: Depending on the interpretation, this might or might not include `0` as valid input.
```suggestion
        ValueError: if value set is not a positive non-zero integer
```

File: metricq/source.py [hunk 2]
```diff
@@ -143,11 +138,21 @@ def task(self):
...
+    def __getitem__(self, id: Metric) -> SourceMetric:
         if id not in self.metrics:
             self.metrics[id] = SourceMetric(id, self, chunk_size=self.chunk_size)
         return self.metrics[id]
 
+    def _augment_metadata(
+        self, metrics: Dict[Metric, MetadataDict]
+    ) -> Dict[Metric, MetadataDict]:
+        augmented: Dict[Metric, MetadataDict] = dict(**metrics)
+        for metric, metadata in augmented.items():
```
Comment (bmario) [inline metricq/source.py:150]: Why is the `augmented` local variable needed? It looks strangely like the input `metrics`.

If for some reason a copy is needed, why not use `copy.copy()` ?

File: tests/test_source.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+class _TestSource(Source):
+    async def task(self):
+        assert False, "This should not be run"
+
+    rpc: AsyncMock
+
+
+@pytest.fixture
+def source():
+
```
Comment (bmario) [inline tests/test_source.py:26]:
```suggestion
```

File: tests/test_source.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
+from sys import version_info
+from typing import Type
+
+if version_info.major == 3 and version_info.minor < 8:
```
Comment (bmario) [inline tests/test_source.py:4]: There's this other PR, which raised the python version to 3.8.

File: setup.cfg [hunk 5]
```diff
@@ -46,6 +46,7 @@ test =
     pytest
     pytest-asyncio
     pytest-mock
+    mock; python_version<'3.8'
```
Comment (bmario) [inline setup.cfg:49]: Is it really only needed for python 3.7 or older? Then good news everyone: #70

**Review by bmario — CHANGES_REQUESTED** [review]
Looks good, but some minor issues

File: tests/test_source.py (hunk 3 above)
Comment (phijor) [reply tests/test_source.py:26]: B-b-b-but... `black` tells me that the empty line is correct :(

File: metricq/source.py (hunk 2 above)
Comment (phijor) [reply metricq/source.py:150]: I'll use `dict.copy()` and add a short comment.

TL;DR: from the outside, `metrics` seems to be a read-only object, so I don't want to have subvert a users expectations and modify the dictionary they own.

File: tests/test_source.py (hunk 4 above)
Comment (phijor) [reply tests/test_source.py:4]: I'll rebase and drop the compat-nonsense :)

File: setup.cfg (hunk 5 above)
Comment (phijor) [reply setup.cfg:49]: Yes, `mock` is the official backport of `unittest.mock`. For `<3.8` it is necessary since `AsyncMock` was not included back then.

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -8,22 +8,23 @@ default_language_version:
 
 repos:
 - repo: https://gitlab.com/pycqa/flake8
-  rev: 3.8.3
+  rev: 3.8.4
```
Comment (bmario) [inline .pre-commit-config.yaml:11]: All changes to this file should probably not be part of this PR

File: metricq/source.py [hunk 7]
```diff
@@ -89,31 +89,26 @@ async def task(self):
 
     """
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.metrics = dict()
-        self._chunk_size: int = 1
+    chunk_size: Optional[int] = cast(Optional[int], ChunkSize())
+    """Number of :term:`data points<Data Point>` collected into a chunk before being sent.
```
Comment (bmario) [inline metricq/source.py:93]: This whole description might not be clear because the chunk size is counted **per** metric. This might leave people with the expression that chunking is done for all metrics combined.

**Review by bmario — CHANGES_REQUESTED** [review]

File: metricq/source.py (hunk 7 above)
Comment (phijor) [reply metricq/source.py]: Does bf31ab8099ae1da5cffb6b2414f144d8ece1be0e make this clear?

File: .pre-commit-config.yaml (hunk 6 above)
Comment (phijor) [reply .pre-commit-config.yaml:11]: Hm, I'll open a separate PR.

**Review by bmario — APPROVED** [review]

**Comment by phijor** [comment]
Found a typo in the last commit... force-pushed it, nothing changed otherwise.

**Comment by bmario** [comment]
@phijor Do you remember the reasoning behind adding the `class ChunkSize()` instead of using a `@property`?

**Comment by phijor** [comment]
Afaik, `ChunkSize` is used in multiple locations (`Source` and `SourceMetric`, as well as tests).  If I remember correctly it's just a "portable" decorator that keeps code DRY by having a single place (`ChunkSize.__set__`) to check the chunk size invariants.
